File: testapp/src/main/java/com/nerdscorner/mvplib/testapp/events/attribute/model/AttributeMainModel.java
package com.nerdscorner.mvplib.testapp.events.attribute.model;

import android.os.AsyncTask;

import com.nerdscorner.mvplib.events.bus.Bus;
import com.nerdscorner.mvplib.events.model.BaseEventsModel;

public class AttributeMainModel extends BaseEventsModel {

    public AttributeMainModel(Bus bus) {
        super(bus);
    }

    public void doSomethingInBackground() {
        new SomeBackgroundTask(this).execute();
    }

    private static class SomeBackgroundTask extends AsyncTask<Void, Void, Void> {
        private AttributeMainModel model;

        SomeBackgroundTask(AttributeMainModel model) {
            this.model = model;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            try {
                Thread.sleep(1000L);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            super.onPostExecute(aVoid);
            model.getBus().post(new BackgroundTaskCompletedEvent());
        }
    }

    public static class BackgroundTaskCompletedEvent {
    }
}

File: testapp/src/main/java/com/nerdscorner/mvplib/testapp/events/ui/mvp/model/AttributeMainModel.kt
package com.nerdscorner.mvplib.testapp.events.ui.mvp.model

import com.nerdscorner.events.coroutines.extensions.runAsync

import com.nerdscorner.mvplib.events.bus.Bus
import com.nerdscorner.mvplib.events.model.BaseEventsModel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class AttributeMainModel(bus: Bus) : BaseEventsModel(bus) {

    fun doSomethingInBackground() {
        CoroutineScope(Dispatchers.Main).launch {
            val deferredResult = runAsync {
                fetchDataAsync()
            }
            val result = deferredResult.await()
            bus.post(BackgroundTaskCompletedEvent(result))
        }
    }

    private suspend fun fetchDataAsync(): String {
        //Fake heavy work
        delay(1000L)
        return "Some important data"
    }

    class BackgroundTaskCompletedEvent(val data: String)
}
